### DISCUSSION TITLE
SMS over IMS

### DISCUSSION THREAD

**Discussion by dushy-3474** [herlesupreeth/docker_open5gs#268] [Q&A]
Hi @herlesupreeth,

We setup a open5gs and Kamailio without docker. After registering the UEs, Volte calls work fine. But SMS is not working.

We are using Samsung m13 commercial phones, so SMS over IMS support should not be an issue. I have even checked Protocol settings ATTACH REQ IEs, MS network support, There is nothing related to SMS over IMS.

There is on for MT SMS over PS, which is coming as 1 in Attach req logs.

Whenever I send an MO SMS, I can see the following error in the Kamailio cconsole?

resolve.c:1740: str2ipbuf() : invalid name: no conversion to IP address possible

Not sure which IP address and which conf file its talking about.

And also in pcrf.log I can see persiodically the following error can be seen.

CHECK WEBUI : Even the DefaultBearer(QCI:9,ARP:8) cannot support IMS signalling.

But as mentioned above, volte calls are working.

Can you please help here.

Thanks
Dushyanth

**Comment by herlesupreeth** [comment]
Please attach a wireshark/tcpdump trace when of scenario where you attach both UEs and then send SMS

Also, even if VoLTE call works there is no guarantee that phone prefers to send SMS over IMS. It may prefer to send SMS over circuit switched network. Its completely upto the settings on the phone

**Reply by dushy-3474** [reply]
Thanks a lot for the reply @herlesupreeth 

I am attaching the pcap capture filtered for SIP and SCTP.
[sms5.zip](https://github.com/herlesupreeth/docker_open5gs/files/13777535/sms5.zip)

**Reply by dushy-3474** [reply]
Note: Please ignore SIP message as some VOIP testing was happening on the same EPC.
Current EPC ip address is 10.0.0.162 and enodeb is 113.

**Reply by herlesupreeth** [reply]
![image](https://github.com/herlesupreeth/docker_open5gs/assets/4115839/6d7f1d5a-6e0f-493e-97f1-60a4578fc4fb)

tel: +7 is not an IMPU registered with your IMS. You would have to send message to 9000000038 or 9000000048

**Reply by dushy-3474** [reply]
Hello Supreeth,

Thanks for the reply.

I am not sure where to add the msisdn.
Its already added to hss_db in FoHSS.
INSERT INTO `impu` VALUES (@impu_id,CONCAT(@sip_str,'@ims.mnc001.mcc001.3gppnetwork.org'),0,1,1,1,@impu_id,1,'','',0,1), (@impu_id+1,CONCAT('tel:',@MSISDN),0,0,0,1,@impu_id+1,1,'','',0,1), (@impu_id+2,CONCAT(@sip_string,'@ims.mnc001.mcc001.3gppnetwork.org'),0,0,1,1,@impu_id,1,'','',0,1);

Something needs to be checked in kamailio.cfg?

Thanks

**Reply by herlesupreeth** [reply]
Please refer to the point **20. Add IMS subscription use in FoHSS as follows from the Web GUI** in the tutorial https://open5gs.org/open5gs/docs/tutorial/02-VoLTE-setup/

**Reply by dushy-3474** [reply]
Hello Supreeth,

Thanks for the email.
I have configured foHSS and HSS with all these numbers using a sql scripts in bulk. Attaching the screenshot of one such number.
Calls are working but and now we know that SMS is going over IMS. But still tel:+7 is going.
I am not able to find out from where it is picking +7. 

I am attaching all the cfgs also. 
Thanks again for the help.
![Screenshot from 2024-01-03 18-34-15](https://github.com/herlesupreeth/docker_open5gs/assets/145524994/ac5464d0-d703-48c6-80a5-8ceb7a797f04)


Regards,
Dushyanth
[configs.tar.gz](https://github.com/herlesupreeth/docker_open5gs/files/13819335/configs.tar.gz)

**Reply by herlesupreeth** [reply]
I see that IMPU tel:9000000038 has Reg. Status as Not-Registered. I believe you have not done correctly the following steps for that

1. Associate IMPI(s) to IMPU
2. Added the IMPU as implicit set of IMSI derived IMPU (i.e. sip:00101xxxxx@ims.xxxxxx)

**Reply by dushy-3474** [reply]
Hello Supreeth,

I did the association with no luck.
![Screenshot from 2024-01-05 11-38-50](https://github.com/herlesupreeth/docker_open5gs/assets/145524994/e00f1f7c-7233-41da-a207-1195d1a2470d)


Regards,
Dushyanth

**Reply by herlesupreeth** [reply]
I would suggest the following:

1. Issue an RTR for all registered IMPUs from the IMPI section for that IMSI
2. Did you go to IMSI derived IMPU and then entered tel:9000000038 under Add IMPU(s) to Implicit-Set and clicked on Add? And not the other around i.e. go to tel:9000000038 IMPU and entered sip:00101xxxxx@ims.xxxxxx under Add IMPU(s) to Implicit-Set and clicked on Add

**Reply by dushy-3474** [reply]
Hi Supreeth,

1st its asking for a reason, I selected new server assigned.
And for the 2nd point. I did both.

**Reply by herlesupreeth** [reply]
> And for the 2nd point. I did both.

you have to do only the first one not both

**Reply by dushy-3474** [reply]
I was trying to follow these points from Volte setup page.


**Login to the HSS web console.
Navigate to the User Identities page
Create the IMSU 
Click IMS Subscription / Create
Enter:
Name = 001010123456791
Capabilities Set = cap_set1
Preferred S-CSCF = scsf1
Click Save

Create the IMPI and Associate the IMPI to the IMSU
Click Create & Bind new IMPI
Enter:
Identity = 001010123456791@ims.mnc001.mcc001.3gppnetwork.org
Secret Key = 8baf473f2f8fd09487cccbd7097c6862 (Ki value as in Open5GS HSS database)
Authentication Schemes - All
Default = Digest-AKAv1-MD5
AMF = 8000 (As in Open5GS HSS database)
OP = 11111111111111111111111111111111 (As in Open5GS HSS database)
SQN = 000000021090 (SQN value as in Open5GS HSS database)
Click Save

Create and Associate IMPI to IMPU
Click Create & Bind new IMPU
Enter:
Identity = sip:001010123456791@ims.mnc001.mcc001.3gppnetwork.org
Barring = Yes
Service Profile = default_sp
Charging-Info Set = default_charging_set
IMPU Type = Public_User_Identity
Click Save

Add Visited Network to IMPU
Enter:
Visited Network = ims.mnc001.mcc001.3gppnetwork.org
Click Add

Now, goto Public User Identity and create further IMPUs as following

1. tel:0198765432100

Public User Identity -IMPU-
Identity = tel:0198765432100
Service Profile = default_sp
Charging-Info Set = default_charging_set
Can Register = Yes
IMPU Type = Public_User_Identity
Click Save

Add Visited Network to IMPU
Enter:
Visited Network = ims.mnc001.mcc001.3gppnetwork.org
Click Add

Associate IMPI(s) to IMPU
IMPI Identity = 001010123456791@ims.mnc001.mcc001.3gppnetwork.org
Click Add

2. sip:0198765432100@ims.mnc001.mcc001.3gppnetwork.org

Public User Identity -IMPU-
Identity = sip:0198765432100@ims.mnc001.mcc001.3gppnetwork.org
Service Profile = default_sp
Charging-Info Set = default_charging_set
Can Register = Yes
IMPU Type = Public_User_Identity
Click Save

Add Visited Network to IMPU
Enter:
Visited Network = ims.mnc001.mcc001.3gppnetwork.org
Click Add

Associate IMPI(s) to IMPU
IMPI Identity = 001010123456791@ims.mnc001.mcc001.3gppnetwork.org
Click Add

And, finally add these IMPUs as implicit set of IMSI derived IMPU in HSS i.e sip:001010123456791@ims.mnc001.mcc001.3gppnetwork.org as follows:

1. Goto to IMPU sip:001010123456791@ims.mnc001.mcc001.3gppnetwork.org
2. Add IMPU(s) to Implicit-Set" section give IMPU Identity created above to be added to this IMPU**

So I added tel:9000000036   and   sip:9000000036@ims.mnc001.mcc001.3gppnetwork.org  to the IMSI derived impu which is 
sip:001010010001036@ims.mnc001.mcc001.3gppnetwork.org  in implicit set and saved. It automatically came in tel:9000000036 's implicit list.
I have saved it. Still tel: is showing as not registered.

**Reply by herlesupreeth** [reply]
If you have followed those steps then please send me a pcap of the registration followed by sending SMS.

**Reply by dushy-3474** [reply]
Hi Supreeth,

Now
I have added 
sip:imsi..@...    impu
Implicit set as tel:msisdn

tel:msisdn  impu
Implicit set as nothing

sip:msisdn 
Implicit set as nothing.

And now tel: is coming as registered and sip:msisdn is coming as not registered.



Still +7 is coming in tel:

Log collected with this setting.
[sms7.zip](https://github.com/herlesupreeth/docker_open5gs/files/13843303/sms7.zip)

**Reply by dushy-3474** [reply]
Hi Supreeth,

After adding tel:mssidn IMPU in the IMSI derived IMPU as implicit set, volte calls stopped working.

The Kamailio logs were showing errors something like cannot convert to ip address.

Does the order in which IMPUs are added to the implicit set matter?

Regards,
Dushyant

**Reply by herlesupreeth** [reply]
> 
> 
> Now
> I have added
> sip:imsi..@... impu
> Implicit set as tel:msisdn

You have to add both tel:msisdn and sip:msisdn@... as implicit set to sip:imsi..@... impu

**Reply by dushy-3474** [reply]
Hi Supreeth,

After doing the changes suggested above, VoLTE calls stopped working. 

As an aside, Would you know what component is responsible for adding "tel:" into the SIP message below (is it scscf, pcscf or icscf, or some block in open5gs/FHoSS)?


#################### START OF MESSAGE #################################
Session Initiation Protocol (MESSAGE)
    **Request-Line: MESSAGE tel:+7 SIP/2.0**
        Method: MESSAGE
        **Request-URI: tel:+7**
        [Resent Packet: False]
    Message Header
        [Expert Info (Warning/Undecoded): Trailing stray characters]
            [Trailing stray characters]
            [Severity level: Warning]
            [Group: Undecoded]
        Via: SIP/2.0/UDP 10.0.0.172:6102;branch=z9hG4bKaa46.941539fb4f963fd9a7f4f6d089e892fa.0;i=71
            Transport: UDP
            Sent-by Address: 10.0.0.172
            Sent-by port: 6102
            Branch: z9hG4bKaa46.941539fb4f963fd9a7f4f6d089e892fa.0
            i=71
        Via: SIP/2.0/TCP 10.76.0.2:6200;received=10.76.0.2;branch=z9hG4bK-524287-1---352b4570b4f37418;rport=6201;transport=TCP
            Transport: TCP
            Sent-by Address: 10.76.0.2
            Sent-by port: 6200
            Received: 10.76.0.2
            Branch: z9hG4bK-524287-1---352b4570b4f37418
            RPort: 6201
            transport=TCP
        Max-Forwards: 69
        Route: <sip:orig@scscf.ims.mnc001.mcc001.3gppnetwork.org:6060;lr>
            Route URI: sip:orig@scscf.ims.mnc001.mcc001.3gppnetwork.org:6060;lr
                Route Userinfo: orig
                Route Host Part: scscf.ims.mnc001.mcc001.3gppnetwork.org
                Route Host Port: 6060
                Route URI parameter: lr
        **To: <tel:+7>**
            **SIP to address: tel:+7**
        From: <tel:9000000036>;tag=1aaa142b
            SIP from address: tel:9000000036
            SIP from tag: 1aaa142b
        Call-ID: kz2uscNqyEZzMlWyYQWZsA..@10.76.0.2
        [Generated Call-ID: kz2uscNqyEZzMlWyYQWZsA..@10.76.0.2]
        CSeq: 1 MESSAGE
            Sequence Number: 1
            Method: MESSAGE
        Allow: MESSAGE
        Content-Type: application/vnd.3gpp.sms
        Supported: path, gruu, sec-agree
        User-Agent: Samsung_SM-M136B_VL1_IMS_6.0
        Request-Disposition: no-fork
        Accept-Contact: *;+g.3gpp.smsip
        P-Access-Network-Info: 3GPP-E-UTRAN-FDD;utran-cell-id-3gpp=40517200010048000
            access-type: 3GPP-E-UTRAN-FDD
            utran-cell-id-3gpp: 40517200010048000
        Content-Length: 22
        P-Charging-Vector: icid-value=4956530A4C0001F24900003200000000; icid-generated-at=10.0.0.172
            icid-value: 4956530A4C0001F24900003200000000
            icid-generated-at=10.0.0.172
        P-Asserted-Identity: <tel:9000000036>
            SIP PAI Address: tel:9000000036
        P-Visited-Network-ID: ims.mnc001.mcc001.3gppnetwork.org
    Message Body
        GSM A-I/F RP - RP-DATA (MS to Network)
            Message Type RP-DATA (MS to Network)
            RP-Message Reference
                RP-Message Reference: 0x10 (16)
            RP-Originator Address
                Length: 0
            RP-Destination Address - (7)
                Length: 2
                1... .... = Extension: No Extension
                .001 .... = Type of number: International Number (0x1)
                .... 0001 = Numbering plan identification: ISDN/Telephony Numbering (ITU-T Rec. E.164 / ITU-T Rec. E.163) (0x1)
                Called Party BCD Number: 7
            RP-User Data
                Length: 15
                TPDU (not displayed)
        GSM SMS TPDU (GSM 03.40) SMS-SUBMIT
            0... .... = TP-RP: TP Reply Path parameter is not set in this SMS SUBMIT/DELIVER
            .0.. .... = TP-UDHI: The TP UD field contains only the short message
            ..0. .... = TP-SRR: A status report is not requested
            ...0 0... = TP-VPF: TP-VP field not present (0)
            .... .0.. = TP-RD: Instruct SC to accept duplicates
            .... ..01 = TP-MTI: SMS-SUBMIT (1)
            TP-MR: 16
            TP-Destination-Address - (9000000032)
                Length: 10 address digits
                1... .... = Extension: No extension
                .000 .... = Type of number: Unknown (0)
                .... 0001 = Numbering plan: ISDN/telephone (E.164/E.163) (1)
                TP-DA Digits: 9000000032
            TP-PID: 0
                00.. .... = Defines formatting for subsequent bits: 0x0
                ..0. .... = Telematic interworking: no telematic interworking, but SME-to-SME protocol
                ...0 0000 = The SM-AL protocol being used between the SME and the MS: 0
            TP-DCS: 0
                00.. .... = Coding Group Bits: General Data Coding indication (0)
                Special case, GSM 7 bit default alphabet
            TP-User-Data-Length: (3) depends on Data-Coding-Scheme
            TP-User-Data
                SMS text: Ho9


#####################

Regards,
Dushyant

**Reply by herlesupreeth** [reply]
If you notice in your pcap, that `tel` is being added by the UE (I think tel:+7 maybe SMSC address programmed in your SIM card). But I noticed that tel:+7 is irrelevant as the SMSC ignores it and only considers the destination address mentioned in SIP body. So I dont think that tel:+7 is the issue.

FHoSS always has been a pain to work with and on top of it even slight misconfiguration will required removing all the subscription details for a SIM and re-provisioning

**Reply by dushy-3474** [reply]
Hello Supreeth,
![Screenshot 2024-01-12 124728](https://github.com/herlesupreeth/docker_open5gs/assets/145524994/7c45a650-f925-4e9e-8d4c-12c49caef716)

Sorry to bug you again. I noticed one thing in the /var/log/syslog
I am attaching the screent shot.
After the i changed the mpath for modules in smsops/examples/kamailio.cfg , Its throwing lot of syntax errors for the uacxxx commands.
Do you have any idea about this?
[kamailio.txt](https://github.com/herlesupreeth/docker_open5gs/files/13915898/kamailio.txt)

Attaching the kamailio.cfg.

The whole file(kamailio.cfg) is same as in kamailio trunk. 
Note: we are starting kamailio with the follwoing.
kamailio -f /usr/local/src/kamailio/src/modules/smsops/examples/kamailio.cfg

Thanks for the help.

**Reply by herlesupreeth** [reply]
hmm, no idea why its throwing that error. But you can use the cfg file from here - https://github.com/herlesupreeth/docker_open5gs/tree/master/smsc
